javascript/76-Minimum-Window-Substring.js

```diff
@@ -1,35 +1,78 @@
-var minWindow = function (str, target) {
-  const hash = target.split('').reduce((acc, val) => {
-    if (!acc[val]) acc[val] = 0;
-    acc[val] += 1;
-    return acc;
-  }, {})
-
-  let start = 0;
-  let min = Infinity;
-  let matched = 0;
-  let subStringStart = null;
-
-  for (let i = 0; i < str.length; i++) {
-    let rightChar = str[i];
-
-    if (rightChar in hash) hash[rightChar] -= 1;
-    if (hash[rightChar] >= 0) matched += 1;
-
-    while (matched === target.length) {
-      if (i - start + 1 < min) {
-        subStringStart = start;
-        min = i - start + 1;
-      }
-
-      let leftChar = str[start];
-      start += 1;
-
-      if (leftChar in hash) {
-        if (hash[leftChar] === 0) matched -= 1;
-        hash[leftChar] += 1;
-      }
+/**
+ * @param {string} s
+ * @param {string} t
+ * Time O(N + M) | SpaceO(N + M)
+ * @return {string}
+ */
+var minWindow = function (s, t) {
+    const isMissingArgs = !s.length || !t.length
+    if (isMissingArgs) return '';
+
+    const frequencyMap = getFrequencyMap(t);
+    const { start, end } = getWindowPointers(s, t, frequencyMap);
+
+    return getSubString(s, start, end);
+}
+
+const getFrequencyMap = (str, frequencyMap = new Map()) => {
+    for (const char of str) {
+        frequencyMap.set(char, (frequencyMap.get(char) || 0) + 1);
+    }
+
+    return frequencyMap;
+};
+
+const getWindowPointers = (s, t, frequencyMap) => {
+    let [ left, right, matched, start, end ] = [ 0, 0, 0, 0, (s.length + 1) ];
+    
+    while (right < s.length) {
+        matched = addRightFrequency(s, right, frequencyMap, matched);
+
+        const canSlide = () => matched === t.length;
+        while (canSlide()) {
+            const window = (right - left) + 1;
+
+            const isSmaller = window < end;
+            if (isSmaller) {
+                [ start, end ] = [ left, window ];
+            }
+
+            matched = subtractLeftFrequency(s, left, frequencyMap, matched);
+            left++;
+        }
+
+        right++;
+    }
+
+    return { start, end };
+};
+
+const addRightFrequency = (s, right, frequencyMap, matched) => {
+    const char = s[right];
+
+    if (frequencyMap.has(char)) {
+        frequencyMap.set(char, frequencyMap.get(char) - 1);
+
+        const isInWindow = 0 <= frequencyMap.get(char);
+        matched += Number(isInWindow);
     }
-  }
-  return min === Infinity ? '' : str.substring(subStringStart, subStringStart + min);
-};
\ No newline at end of file
+
+    return matched;
+};
+
+const subtractLeftFrequency = (s, left, frequencyMap, matched) => {
+    const char = s[left];
+
+    if (frequencyMap.has(char)) {
+        const isOutOfWindow = frequencyMap.get(char) === 0;
+
+        matched -= Number(isOutOfWindow);
```

Comment:
for readability's sake, converting a boolean to a number feels potentially confusing. maybe prefer:
```
if (isOutOfWindow) {
  matched--
}
```